File: src/windows-native/LanMessenger/Core/Protocol/FrameCodec.cs
using System.Buffers.Binary;
using System.Text;
using System.Text.Json;

namespace LanMessenger.Core.Protocol;

public static class FrameCodec
{
    public const int MaxFrameSize = 50 * 1024 * 1024; // 50 MiB

    private static readonly JsonSerializerOptions _compact = new()
    {
        WriteIndented = false,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };

    // Encodes a typed object as a length-prefixed frame.
    // Layout: [4 bytes big-endian uint32 length][UTF-8 JSON body]
    public static byte[] Encode<T>(T value)
    {
        byte[] body = JsonSerializer.SerializeToUtf8Bytes(value, _compact);
        return BuildFrame(body);
    }

    // Encodes a raw dictionary as a length-prefixed frame.
    public static byte[] EncodeDict(Dictionary<string, object?> dict)
    {
        byte[] body = JsonSerializer.SerializeToUtf8Bytes(dict, _compact);
        return BuildFrame(body);
    }

    // Reads exactly one frame from stream. Returns null on clean EOF.
    // Throws IOException on protocol violations.
    public static byte[]? ReadFrame(Stream stream)
    {
        byte[]? header = ReadExact(stream, 4);
        if (header is null) return null;

        int length = (int)BinaryPrimitives.ReadUInt32BigEndian(header);
        if (length <= 0 || length > MaxFrameSize)
            throw new InvalidDataException($"Invalid frame size: {length}");

        byte[]? body = ReadExact(stream, length);
        return body; // null means connection closed mid-frame (treated as error by caller)
    }

    // Async version for use with NetworkStream.
    public static async Task<byte[]?> ReadFrameAsync(Stream stream, CancellationToken ct = default)
    {
        byte[]? header = await ReadExactAsync(stream, 4, ct).ConfigureAwait(false);
        if (header is null) return null;

        int length = (int)BinaryPrimitives.ReadUInt32BigEndian(header);
        if (length <= 0 || length > MaxFrameSize)
            throw new InvalidDataException($"Invalid frame size: {length}");

        return await ReadExactAsync(stream, length, ct).ConfigureAwait(false);
    }

    // Parses a frame body to a JsonDocument.
    public static JsonDocument ParseJson(byte[] data) => JsonDocument.Parse(data);

    // MARK: - Helpers

    private static byte[] BuildFrame(byte[] body)
    {
        if (body.Length == 0 || body.Length > MaxFrameSize)
            throw new InvalidOperationException($"Frame body out of range: {body.Length}");
        byte[] frame = new byte[4 + body.Length];
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(0, 4), (uint)body.Length);
        body.CopyTo(frame, 4);
        return frame;
    }

    // Returns null on clean EOF at the start; throws on mid-read disconnect.
    private static byte[]? ReadExact(Stream stream, int count)
    {
        byte[] buf = new byte[count];
        int total = 0;
        while (total < count)
        {
            int n = stream.Read(buf, total, count - total);
            if (n == 0)
            {
                if (total == 0) return null; // clean EOF
                throw new IOException("Connection closed mid-frame");
            }
            total += n;
        }
        return buf;
    }

    private static async Task<byte[]?> ReadExactAsync(Stream stream, int count, CancellationToken ct)
    {
        byte[] buf = new byte[count];
        int total = 0;
        while (total < count)
        {
            int n = await stream.ReadAsync(buf.AsMemory(total, count - total), ct).ConfigureAwait(false);
            if (n == 0)
            {
                if (total == 0) return null;
                throw new IOException("Connection closed mid-frame");
            }
            total += n;
        }
        return buf;
    }
}

File: src/macos/LanMessenger/Core/Protocol/FrameCodec.swift
import Foundation

enum FrameCodecError: Error {
    case frameTooLarge(Int)
    case invalidFrameSize
    case connectionClosed
    case encodingFailed
}

enum FrameCodec {

    static let maxFrameSize = 50 * 1024 * 1024  // 50 MiB

    // Encode a Codable value as a length-prefixed frame.
    // Layout: [4 bytes big-endian uint32 length][UTF-8 JSON body]
    static func encode<T: Encodable>(_ value: T) throws -> Data {
        let body = try JSONEncoder().encode(value)
        guard body.count > 0 && body.count <= maxFrameSize else {
            throw FrameCodecError.frameTooLarge(body.count)
        }
        var frame = Data(capacity: 4 + body.count)
        var length = UInt32(body.count).bigEndian
        frame.append(Data(bytes: &length, count: 4))
        frame.append(body)
        return frame
    }

    // Encode a raw [String: Any] dict as a length-prefixed frame.
    static func encodeDict(_ dict: [String: Any]) throws -> Data {
        let body = try JSONSerialization.data(withJSONObject: dict)
        guard body.count > 0 && body.count <= maxFrameSize else {
            throw FrameCodecError.frameTooLarge(body.count)
        }
        var frame = Data(capacity: 4 + body.count)
        var length = UInt32(body.count).bigEndian
        frame.append(Data(bytes: &length, count: 4))
        frame.append(body)
        return frame
    }

    // Read one frame from a stream. Returns nil on clean close.
    // Throws FrameCodecError on protocol violations.
    static func readFrame(from stream: InputStream) throws -> Data? {
        // Read 4-byte header
        var header = [UInt8](repeating: 0, count: 4)
        let headerRead = try readExact(stream: stream, buffer: &header, count: 4)
        guard headerRead else { return nil }   // clean EOF

        let length = Int(UInt32(bigEndian: header.withUnsafeBytes { $0.load(as: UInt32.self) }))
        guard length > 0 && length <= maxFrameSize else {
            throw FrameCodecError.frameTooLarge(length)
        }

        var body = [UInt8](repeating: 0, count: length)
        guard try readExact(stream: stream, buffer: &body, count: length) else {
            return nil
        }
        return Data(body)
    }

    // Parse a frame body into a [String: Any] dict.
    static func parseJSON(from data: Data) throws -> [String: Any] {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw FrameCodecError.encodingFailed
        }
        return dict
    }

    // MARK: - Private helpers

    private static func readExact(stream: InputStream, buffer: inout [UInt8], count: Int) throws -> Bool {
        var totalRead = 0
        while totalRead < count {
            let n = buffer.withUnsafeMutableBytes { ptr in
                stream.read(ptr.baseAddress!.advanced(by: totalRead).assumingMemoryBound(to: UInt8.self),
                            maxLength: count - totalRead)
            }
            if n < 0 { throw FrameCodecError.connectionClosed }
            if n == 0 {
                if totalRead == 0 { return false }
                throw FrameCodecError.connectionClosed
            }
            totalRead += n
        }
        return true
    }
}
